README.md
![Clay Containers for implementing beautiful, modern neumorphic designs.](https://res.cloudinary.com/mca62511/image/upload/v1579847668/banner_zx6prd.png)

# Clay Containers

Easily create and customize beautiful, modern [neumorphic](https://dribbble.com/tags/neumorphism) containers for your Flutter project. These clay containers can become the basis for your own unique neumorphic designs.

## Installation

Add `clay_containers` to your project as a [dependency in your pubspec.yaml file](https://flutter.dev/docs/development/packages-and-plugins/using-packages). This is a simple Dart plugin, so additional configuration for iOS and Android is not needed.

## Examples

### Simple `ClayContainer`

For best results, set the background
color of a surrounding widget to match
the color you will set for your clay
container. Since it is likely you'll reuse this base color
multiple times (especially if you end up doing something fancy)
it's good to set this color to a single value. In the following example it
is set to `baseColor`.

```
import 'package:clay_containers/clay_containers.dart';

class MyExampleScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    Color baseColor = Color(0xFFF2F2F2);

    return Container(
        color: baseColor,
        child: Center(
          child: ClayContainer(
            color: baseColor,
            height: 200,
            width: 200,
          ),
        ),
      );
  }
}
```

![ClayContainer example.](https://res.cloudinary.com/mca62511/image/upload/v1579847714/simple_xeh3pd.png)

### `ClayContainer` with a `ClayText` Child.

In the previous example the `ClayContainer` was given `height`
and `width` since it has no child. `ClayContainer` behaves the
same as a normal `Container` and needs to be either given
`height` and `width` or a `child` to be visible. In the
following example, the `ClayContainer` will receive a child.

The child it will receive is a `ClayText` wrapped in some `Padding`.

```
ClayContainer(
          color: baseColor,
          child: Padding(
            padding: EdgeInsets.all(20),
            child: ClayText("Seize the Clay!", emboss: true, size: 40),
          ),
        ),
```

![Clay container example with child.](https://res.cloudinary.com/mca62511/image/upload/v1579847742/simple_child_text_jmwjd3.png)

### Rounded `ClayContainer`s

Don't be a square! Use borderRadius to add some flare. If you want a uniform `borderRadius` you can simply set it directly in the `ClayContainer` constructor.

```
ClayContainer(
          color: baseColor,
          height: 150,
          width: 150,
          borderRadius: 50,
        ),
```

![A rounded ClayContainer.](https://res.cloudinary.com/mca62511/image/upload/v1579847766/circle_ci1at9.png)

If you want to pass your own custom `BorderRadius` object, that is available as well: In that case pass it to `customBorderRadius`.

```
ClayContainer(
          color: baseColor,
          height: 150,
          width: 150,
          customBorderRadius: BorderRadius.only(
              topRight: Radius.elliptical(150, 150),
              bottomLeft: Radius.circular(50)),
        ),
```

![A weird shaped ClayContainer.](https://res.cloudinary.com/mca62511/image/upload/v1579847787/weird_pv8vnk.png)

### Embossed `ClayContainer`s

You may have noticed earlier that the `ClayText` can receive an `emboss` property. ClayContainers can as well. All clay widgets start in a debossed state by default.

```
ClayContainer(
          emboss: true,
          color: baseColor,
          height: 150,
          width: 150,
          borderRadius: 50,
        ),
```

![An embossed ClayContainer.](https://res.cloudinary.com/mca62511/image/upload/c_scale,w_570/v1579930865/ss__2020-01-25_at_14.34.08_hntksj.png)

### Change Default Spread and Depth

Don't like the default look of the neumorphic effect? Change the base variables. Do whatever you want. I'm not your mom.

```
ClayContainer(
          color: baseColor,
          height: 150,
          width: 150,
          borderRadius: 75,
          depth: 40,
          spread: 40,
        ),
```

![Embossed!](https://res.cloudinary.com/mca62511/image/upload/v1579847841/deep_v010zd.png)

### Concave and Convex `ClayContainer`s

Give your `ClayContainer` a convex or a concave look by passing either `CurveType.concave` or `CurveType.convex` to the `curveType` parameter.

```
Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            ClayContainer(
              color: baseColor,
              height: 150,
              width: 150,
              borderRadius: 75,
              curveType: CurveType.concave,
            ),
            SizedBox(width: 50),
            ClayContainer(
              color: baseColor,
              height: 150,
              width: 150,
              borderRadius: 75,
              curveType: CurveType.none,
            ),
            SizedBox(width: 50),
            ClayContainer(
              color: baseColor,
              height: 150,
              width: 150,
              borderRadius: 75,
              curveType: CurveType.convex,
            ),
          ],
        ),
```

![Concave, flat, and convex ClayContainers.](https://res.cloudinary.com/mca62511/image/upload/v1579847860/concave_convex_hyafpx.png)

### Animate It!

You can use a `ClayAnimatedContainer` to create animations in the same way as you would an `AnimatedContainer`. An explanation of `AnimatedContainer`s can be found on the [Google Developers channel](https://www.youtube.com/watch?v=yI-8QHpGIP4) on YouTube.

![Very animated. Much cool.](https://res.cloudinary.com/mca62511/image/upload/c_scale,h_300,w_300/v1579847878/animated_rktpdo.gif)



### Theming

You asked for it, you got it. Now you can specify your Clay Container's properties high up in your widget tree and that theme will apply to every Clay container you create under that `ClayTheme`. Simply introduce the `ClayTheme` anywhere in your widget tree and pass it a `ClayThemeData` to it's `themeData` property

```dart
ClayTheme(
  themeData: const ClayThemeData(
    height: 10,
    width: 20,
    borderRadius: 360,
    textTheme: ClayTextTheme(style: TextStyle()),
    depth: 12,
  ),
  child: (...),
);
```

Now any `ClayWidget` that is an ancestor of this `ClayTheme` will inherit it's theme.

Also notice the `ClayTextTheme` which will handle the theme of your `ClayText`.

You can override any of these properties to give a specific `ClayWidget` custom theming. Simply give that property a value when you create it.

```dart
ClayTheme(
  themeData: const ClayThemeData(
    height: 10,
    width: 20,
    borderRadius: 360,
    textTheme: ClayTextTheme(style: TextStyle()),
    depth: 12,
  ),
  child: ClayAnimatedContainer(
      height: 240,
      width: 240,
      child: ClayContainer(
          borderRadius: 10,
      ),
  ),
);
```

Now, the `ClayAnimatedContainer` in the code above will have a `height` and `width` of `240`, ignoring the `10` and `20` height and width respectively passed to the `ClayTheme`, but the `ClayContainer` will rather take `height` and `width` of `10` and `20` respectively, which is from the `ClayTheme`, but will override the `borderRadius` with a `10`.

Both the `ClayThemeData` and the `ClayTextTheme` have only one default value

```dart
color: const Color(0xFFf0f0f0);
```



### Full API Documentation

#### `ClayContainer`

##### Positional Arguments

None.

##### Named Arguments

- **color** - This sets the base color for the clay object. Simply setting this to the background color of the parent object will give you a pretty good looking debossed neumorphic effect.
- **height** - This sets the height of the container.
- **width** - This sets the width of the container.
- **parentColor** - This tells the widget to use a different color for the outside emboss/deboss effect, despite whatever is set in the `color` field.
- **surfaceColor** - This tells the widget to use a different color for the inside of the container, despite whatever is set in the `color` field.
- **spread** - How far should the emboss/deboss effect spread?
- **depth** - How strong should the emboss/deboss effect be?
- **child** - This receives child widgets.
- **borderRadius** - This receives a number representing a border radius to be applied to all corners of the container.
- **customBorderRadius** - This receives a `BorderRadius` object. Setting this object will override whatever is set in the `borderRadius` field.
- **curveType** - This receives a `CurveType` enum. Use this to set the inside surface to look either convex or concave.
- **emboss** - This is `false` by default. Set this to `true` in order to make an embossed container.

#### `ClayText`

##### Positional Arguments

- **text** - This is the text to be displayed.

##### Named Arguments

- **color** - This sets the base color for the clay object. Simply setting this to the background color of the parent object will give you a pretty good looking debossed neumorphic effect.
- **parentColor** - This tells the widget to use a different color for the outside emboss/deboss effect, despite whatever is set in the `color` field.
- **textColor** - This tells the widget to use a different color for the fill of the text, despite whatever is set in the `color` field.
- **spread** - How far should the emboss/deboss effect spread?
- **depth** - How strong should the emboss/deboss effect be?
- **emboss** - This is `false` by default. Set this to `true` in order to make an embossed container.

#### `ClayAnimatedContainer`

##### Positional Arguments

None.

##### Named Arguments

- **duration** - The duration over which to animate the parameters of this container.
- **curve** - The curve to apply when animating the parameters of this container.
- **Additional parameters are identical to `ClayContainer`.**


CHANGELOG.md
## [0.3.4] - December 27, 2023

- Feature: ClayTheme - Thanks to [NonymousMorlock](https://github.com/NonymousMorlock)
- Feature: ClayAnimatedContainer onEnd implementation - Thanks to [NonymousMorlock](https://github.com/NonymousMorlock)

## [0.3.3] - March 9, 2021

- Hotfix: Same as 0.3.1, but for ClayAnimatedContainer.

## [0.3.2] - March 9, 2021

- Hotfix: Fixed problem where clay containers without a curve would throw an error.

## [0.3.0] - March 9, 2021

- Null Safety! Thanks to [Sameerkash](https://github.com/Sameerkash)

## [0.2.3] - January 30, 2021

- Added ClayAnimatedContainer thanks to [babakoto](https://github.com/babakoto)

## [0.2.2] - January 25, 2020

- Yet another minor documentation adjustment.

## [0.2.1] - January 25, 2020

- Made default embossed surfaces brighter.

## [0.2.0] - January 24, 2020

- Introduce the `ClayText` widget.
- Add the `emboss` property to `ClayContainer`s.
- Allow arbitrarily setting `surfaceColor` and `parentColor` for `ClayContainer`s.
- `ClayContainer`s no longer require a color, but will instead default to `Color(0xFFF0F0F0)`.
- Documentation updates.

## [0.1.12] - January 22, 2020

- Documentation changes.

## [0.1.11] - January 22, 2020

- Documentation changes.
- Removed unused variable.

## [0.1.2] - January 24, 2020

- This release was a mistake. 0.1.2 is meant to be the release after 0.1.12. I misunderstood the way the version numbering works.

## [0.1.1] - January 21, 2020

- Initial release.


example/lib/main.dart
import 'package:clay_containers/clay_containers.dart';
import 'package:flutter/material.dart';

void main() => runApp(const MaterialApp(home: MyApp()));

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with SingleTickerProviderStateMixin {
  Color baseColor = const Color(0xFFf2f2f2);
  double firstDepth = 50;
  double secondDepth = 50;
  double thirdDepth = 50;
  double fourthDepth = 50;
  late AnimationController _animationController;

  @override
  void initState() {
    _animationController = AnimationController(
      vsync: this, // the SingleTickerProviderStateMixin
      duration: const Duration(seconds: 5),
    )..addListener(() {
        setState(() {});
      });

    _animationController.forward();

    super.initState();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    double? stagger(double value, double progress, double delay) {
      var newProgress = progress - (1 - delay);
      if (newProgress < 0) newProgress = 0;
      return value * (newProgress / delay);
    }

    final calculatedFirstDepth =
        stagger(firstDepth, _animationController.value, 0.25)!;
    final calculatedSecondDepth =
        stagger(secondDepth, _animationController.value, 0.5)!;
    final calculatedThirdDepth =
        stagger(thirdDepth, _animationController.value, 0.75)!;
    final calculatedFourthDepth =
        stagger(fourthDepth, _animationController.value, 1)!;

    return ClayTheme(
      themeData: const ClayThemeData(
        height: 10,
        width: 20,
        borderRadius: 360,
        textTheme: ClayTextTheme(style: TextStyle()),
        depth: 12,
      ),
      child: ColoredBox(
        color: baseColor,
        child: Center(
          child: ClayContainer(
            color: baseColor,
            height: 240,
            width: 240,
            curveType: CurveType.concave,
            spread: 30,
            depth: calculatedFirstDepth.toInt(),
            child: Center(
              child: ClayContainer(
                height: 200,
                width: 200,
                depth: calculatedSecondDepth.toInt(),
                curveType: CurveType.convex,
                color: baseColor,
                child: Center(
                  child: ClayContainer(
                    height: 160,
                    width: 160,
                    color: baseColor,
                    depth: calculatedThirdDepth.toInt(),
                    curveType: CurveType.concave,
                    child: Center(
                      child: ClayContainer(
                        height: 120,
                        width: 120,
                        color: baseColor,
                        depth: calculatedFourthDepth.toInt(),
                        curveType: CurveType.convex,
                      ),
                    ),
                  ),
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
